core/src/main/java/bisq/core/btc/TxFeeEstimationService.java

```diff
@@ -0,0 +1,159 @@
+                preferences);
+    }
+
+    public Tuple2<Coin, Integer> getEstimatedFeeAndTxSizeForMaker(Coin reservedFundsForOffer,
+                                                                  Coin tradeFee) {
+        return getEstimatedFeeAndTxSize(false,
+                reservedFundsForOffer,
+                tradeFee,
+                feeService,
+                btcWalletService,
+                preferences);
+    }
+
+    private Tuple2<Coin, Integer> getEstimatedFeeAndTxSize(boolean isTaker,
+                                                           Coin amount,
+                                                           Coin tradeFee,
+                                                           FeeService feeService,
+                                                           BtcWalletService btcWalletService,
+                                                           Preferences preferences) {
+        Coin txFeePerByte = feeService.getTxFeePerByte();
+        // We start with min taker fee size of 260
+        int estimatedTxSize = 260;
+        try {
+            estimatedTxSize = getEstimatedTxSize(List.of(tradeFee, amount), estimatedTxSize, txFeePerByte, btcWalletService);
+        } catch (InsufficientMoneyException e) {
+            if (isTaker) {
+                // if we cannot do the estimation we use the payout tx size
+                estimatedTxSize = 380;
+            }
+            log.info("We cannot do the fee estimation because there are not enough funds in the wallet. This is expected " +
+                    "if the user pays from an external wallet. In that case we use an estimated tx size of {} bytes.", estimatedTxSize);
+        }
+
+        if (!preferences.isPayFeeInBtc()) {
+            // If we pay the fee in BSQ we have one input more which adds about 150 bytes
+            estimatedTxSize += 150;
+        }
+
+        Coin txFee;
+        int size;
+        if (isTaker) {
+            int averageSize = (estimatedTxSize + 320) / 2;  // deposit tx has about 320 bytes
+            // We use at least the size of the payout tx to not underpay at payout.
+            size = Math.max(380, averageSize);
+            txFee = txFeePerByte.multiply(size);
+            log.info("Fee estimation resulted in a tx size of {} bytes.\n" +
+                    "We use an average between the taker fee tx and the deposit tx (320 bytes) which results in {} bytes.\n" +
+                    "The payout tx has 380 bytes, we use that as our min value. Size for fee calculation is {} bytes.\n" +
+                    "The tx fee of {} Sat", estimatedTxSize, averageSize, size, txFee.value);
+        } else {
+            size = estimatedTxSize;
+            txFee = txFeePerByte.multiply(size);
+            log.info("Fee estimation resulted in a tx size of {} bytes and a tx fee of {} Sat.", size, txFee.value);
+        }
+
+        return new Tuple2<>(txFee, size);
+    }
+
+    @VisibleForTesting
```

Comment:
I've been always told that this is a bad idea, as it's an attempt to test implementation details.
Either extract it to an utility class that could be tested separately or do not test it individually.

Reply:
Not clear what you mean. Lets discuss on a call in a bit...

Reply:
Ah ok. Yes basically agree. In that case I would not like to expose getEstimatedTxSize as it is only interesting for that class. The more high level methods carry much more special case handling that they would be harder to test then the core functionality in getEstimatedTxSize.